In [1]:
// подключаем необходимые библиотеки
import $ivy.`org.scalanlp::breeze:1.2`
import scala.collection.mutable.ArrayBuffer
import breeze.linalg._

import $ivy.$                         

import scala.collection.mutable.ArrayBuffer

import breeze.linalg._

In [2]:
// читаем файл с данными
val dataFile = io.Source.fromFile("Dataset.csv")

// подготавливаем матрицу
var dataMatrix = DenseMatrix.zeros[Double](0,28)

var i = 0
// Пробегаемся по каждай строке файла
for (line <- dataFile.getLines.drop(1)) { 
    
    // готовим массив для дальнейшего его перегона в матрицу 
    var dataArray = Array[Double]()
    
    // делим строку на элементы и обрезаем пробелы
    for (elem <- line.split(",")map(_.trim)){
        // если значения нет, то добавляем 0
        if(elem == "") {
            dataArray = dataArray :+ 0.0 
        // иначе добавляем в массив значение елемента
        }else{
            dataArray = dataArray :+ elem.toDouble
        }
    }
    
    // наполняем матрицу
    dataMatrix = DenseMatrix.vertcat(dataMatrix, DenseMatrix(DenseVector(dataArray)))
    i += 1 // без счетчика у меня напроч умирал юпитер, видимо потому, что пытался вывести все значения матрицы 
}



dataFile: io.BufferedSource = empty iterator
dataMatrix: DenseMatrix[Double] = 634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    65.0  ... (28 total)
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    10.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    14.0  ...
634995.0  0.0  463.0  1.0  7.0   0.0   3.0   7.0   -3.0   62.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    58.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    60.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    68.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   32.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    35.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    48.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    52.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    3.0   ...
634995.0  0.0  463.0  1.0  1.0   1.0   0.0   1.0   1.0    37.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    23.0  ...
634995.0  0.0  463.0  1.0  3.0   0.0   3.0   3.0   -3.0   40.0  ...
634995.0  0.0  463.0  1.0  3.0   0.0   3.0   2.0   -3.0   54.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    29.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   36.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    11.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    62.0  ...
634995.0  0.0  463.0  1.0  77.0  46.0  31.0  74.0  15.0   31.0  ...
634995.0  0.0  463.0  1.0  6.0   0.0   1.0   6.0   -1.0   54.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    71.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    61.0  ...
634995.0  0.0  463.0  1.0  4.0   4.0   0.0   4.0   4.0    13.0  ...
634995.0  0.0  463.0  1.0  29.0  3.0   11.0  25.0  -8.0   49.0  ...
634995.0  0.0  463.0  1.0  3.0   1.0   2.0   3.0   -1.0   35.0  ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    11.0  ...
634995.0  0.0  463.0  1.0  2.0   0.0   2.0   2.0   -2.0   37.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    0.0   ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    11.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    43.0  ...
634995.0  0.0  463.0  1.0  2.0   2.0   0.0   2.0   2.0    18.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    64.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    51.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    67.0  ...
634995.0  0.0  463.0  1.0  8.0   8.0   0.0   8.0   8.0    12.0  ...
...
i: Int = 40949

In [3]:
// разбиваем данные
val X_all = dataMatrix(0 to dataMatrix.rows - 1, 0 to dataMatrix.cols-2)
val Y_all = dataMatrix(::,27)

X_all: DenseMatrix[Double] = 634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    65.0  ... (27 total)
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    10.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    14.0  ...
634995.0  0.0  463.0  1.0  7.0   0.0   3.0   7.0   -3.0   62.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    58.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    60.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    68.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   1.0   1.0   -1.0   32.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    35.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    48.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    52.0  ...
634995.0  0.0  463.0  1.0  1.0   0.0   0.0   1.0   0.0    69.0  ...
634995.0  0.0  463.0  1.0  0.0   0.0   0.0   0.0   0.0    3.0   ...
634995.0  0.0  463.0  1.0  1.0   1.0   0.0   1.0   1.0    37.0  ...
634995.0

In [4]:
// Посмотрим минимальное, максимальное и среднее значения целевого признака

import breeze.stats.mean
println("Min value = " + Y_all.min)
println("Max value = " + Y_all.max)
println("Mean value = " + mean(Y_all))


Min value = 0.0
Max value = 1305.0
Mean value = 7.322889447849774


import breeze.stats.mean


**не осилил это сделать в скале, сделал в питоновском ноуте и вставил скринами**


#### Посмотрим на распределение 
![title](distrib.png)


#### Посмотрим на корреляцию
![title](dataset_corr.png)

Как мы можем заметить, самая высока степерь корреляции с целевым признаком output у признака **comm24**. Далее по списку идут: **diff2448 , comm24_1 , commBase**. Их и будем использовать в качестве описываемых признаков.

In [5]:
// вырезаем необходимые признаки (4,5,7,8), ормализуем их и собираем новую матрцу
import breeze.optimize._

val X_commBase = DenseMatrix(X_all(::,4))  
val X_comm24 = DenseMatrix(X_all(::,5)) 
val X_comm24_1 = DenseMatrix(X_all(::,7)) 
val X_diff2448 = DenseMatrix(X_all(::,8)) 
var X = DenseMatrix.zeros[Double](X_commBase(::,0).length,0)

X = DenseMatrix.horzcat(X_commBase.t, X_comm24.t)
X = DenseMatrix.horzcat(X, X_comm24_1.t)
X = DenseMatrix.horzcat(X, X_diff2448.t)

import breeze.optimize._


X_commBase: DenseMatrix[Double] = 0.0  0.0  0.0  7.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  ... (40949 total)
X_comm24: DenseMatrix[Double] = 0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ... (40949 total)
X_comm24_1: DenseMatrix[Double] = 0.0  0.0  0.0  7.0  1.0  0.0  0.0  1.0  0.0  0.0  0.0  1.0  ... (40949 total)
X_diff2448: DenseMatrix[Double] = 0.0  0.0  0.0  -3.0  0.0  0.0  0.0  -1.0  0.0  0.0  0.0  0.0  ... (40949 total)
X: DenseMatrix[Double] = 0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
7.0   0.0   7.0   -3.0   
1.0   0.0   1.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   -1.0   
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   0.0    
0.0   0.0   0.0   0.0    
1.0   1.0   1.0   1.0    
0.0   0.0   0.0   0.0    
3.0   0.0   3.0   -3.0   
3.0   0.0   2.0   -3.0   
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   -1.0   
0.0   

In [6]:
// разделим выборки на train и test
val test_count = (X.rows * 0.67).toInt
val X_train = X(0 to test_count, 0 to X.cols - 1)
val X_test = X(test_count to X.rows - 1, 0 to X.cols - 1)
val Y_train = Y_all(0 to test_count)
val Y_test = Y_all(test_count to Y_all.length - 1)
assert(X_train.rows == Y_train.length)
assert(X_test.rows == Y_test.length)

test_count: Int = 27435
X_train: DenseMatrix[Double] = 0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
7.0   0.0   7.0   -3.0   
1.0   0.0   1.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   -1.0   
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   0.0    
0.0   0.0   0.0   0.0    
1.0   1.0   1.0   1.0    
0.0   0.0   0.0   0.0    
3.0   0.0   3.0   -3.0   
3.0   0.0   2.0   -3.0   
0.0   0.0   0.0   0.0    
1.0   0.0   1.0   -1.0   
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
77.0  46.0  74.0  15.0   
6.0   0.0   6.0   -1.0   
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
0.0   0.0   0.0   0.0    
4.0   4.0   4.0   4.0    
29.0  3.0   25.0  -8.0   
3.0   1.0   3.0   -1.0   
2.0   2.0   2.0   2.0    
2.0   0.0   2.0   -2.0   
0.0   0.0   0.0   0.0    
2.0   2.0   2.0   2.0    
0.0   0.0   0.0   0.0    
2.0   2.0   2.0   2.0    
0.0   0.0   0.0   0.0    
0.0   0.0

In [7]:
// применяем регрессионные модели на 4х колонках
import breeze.stats.regression.{leastSquares, lasso}
val squaresModel = leastSquares(X_train, Y_train)
val lassoModel = lasso(X_train, Y_train, 0.01)


Apr 15, 2021 8:57:03 PM com.github.fommil.netlib.LAPACK <clinit>
Apr 15, 2021 8:57:03 PM com.github.fommil.netlib.LAPACK <clinit>


import breeze.stats.regression.{leastSquares, lasso}

squaresModel: breeze.stats.regression.LeastSquaresRegressionResult = LeastSquaresRegressionResult(
  DenseVector(0.13764243010575908, 0.27411510852536086, -0.15292795367722417, 0.016967134234060353),
  2.515324777289501E7
)
lassoModel: breeze.stats.regression.LassoResult = LassoResult(
  DenseVector(0.0, 0.2551181303332685, 0.0, 0.009294770197934009),
  2.5226542751696546E7,
  0.01
)

In [8]:
// применим наши модели на тестовой выборке

assert(X_test.cols == squaresModel.coefficients.length)
val Y_pred_squaresModel = X_test * squaresModel.coefficients
val Y_pred_lassoModel = X_test * lassoModel.coefficients


Apr 15, 2021 8:57:04 PM com.github.fommil.netlib.BLAS <clinit>
Apr 15, 2021 8:57:04 PM com.github.fommil.netlib.BLAS <clinit>


Y_pred_squaresModel: DenseVector[Double] = DenseVector(-4.023403509631349, 0.8388782378772148, 9.652885171578465, -6.497753453037294, 74.46511418074816, 5.196706061385847, -6.64953756955302, 277.1757027838959, 0.0, 0.6165695020864126, -2.7866353488156435, -0.5681954199688392, -0.3275636577236781, -1.1763781537077378, 1.0803743403396489, 1.2532712203996583, 2.17756687314776, -0.3838196999492224, 11.307665486706199, 21.272644292421262, 0.606098800647069, -0.05066215249277395, 28.13126535717153, -2.34098122154683, 0.011995020208847844, 26.752281761231746, 0.34842413534262695, 19.029973623968967, 33.837298676982144, 11.583462205894158, 2.0183656344383696, 18.47838018559306, 3.5853573494434303, -0.509446013524447, 51.5739864881478, -3.341538034699316, -0.2953109999181521, -0.6826533737220604, 38.61154068631386, 27.209131303539202, -0.5127442692064268, 0.0188754665360904, -3.742028405349345, 14.89302283614963, 5.240137664571167, 13.10336527745011, 0.03884916916948494, 68.39758635861311, 35.8

### В качестве функции потерь будем использовать MAE т.к. нам необходимо понимать, на сколько комментариев мы промахиваемся в ту или иную сторону

In [9]:
import math.abs
val mae = (X: DenseVector[Double], X_pred: DenseVector[Double]) => {
    ((X - X_pred).map(x => abs(x)).sum) / X.length
}

val mae_test_squaresModel = mae(Y_test, Y_pred_squaresModel)
val mae_test_lassoModel = mae(Y_test, Y_pred_lassoModel)


println(mae_test_squaresModel)
println(mae_test_lassoModel)

6.231902929890508
5.984693119061619


import math.abs

mae: (DenseVector[Double], DenseVector[Double]) => Double = ammonite.$sess.cmd8$Helper$$Lambda$2723/2145321045@46c5b153
mae_test_squaresModel: Double = 6.231902929890508
mae_test_lassoModel: Double = 5.984693119061619

##### Судя по результатам оценки, lassoModel показывает немного большую точность чем squaresModel на текущих данных

## Вычислим коэфиценты своими силами и сравним результаты

### Т.к. данных у нас не так много, можно использовать самую простую реализацию

In [10]:
import breeze.linalg.inv
val X_train_t = X_train.t
val coef_custom = inv((X_train_t * X_train)) * X_train_t * Y_train
var Y_pred_custom = X_test * coef_custom
val mae_test_custom = mae(Y_test, Y_pred_custom)

println(mae_test_custom)

6.231902929890598


import breeze.linalg.inv

X_train_t: DenseMatrix[Double] = 0.0  0.0  0.0  7.0   1.0  0.0  0.0  1.0   0.0  0.0  0.0  1.0  ... (27436 total)
0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0   0.0  0.0  0.0  0.0  ...
0.0  0.0  0.0  7.0   1.0  0.0  0.0  1.0   0.0  0.0  0.0  1.0  ...
0.0  0.0  0.0  -3.0  0.0  0.0  0.0  -1.0  0.0  0.0  0.0  0.0  ...
coef_custom: DenseVector[Double] = DenseVector(0.13764243010575952, 0.2741151085253701, -0.15292795367722656, 0.016967134234060682)
Y_pred_custom: DenseVector[Double] = DenseVector(-4.023403509631688, 0.8388782378770561, 9.65288517157873, -6.497753453038027, 74.4651141807502, 5.196706061385955, -6.649537569554052, 277.1757027839035, 0.0, 0.61656950208632, -2.786635348815977, -0.5681954199690411, -0.3275636577237324, -1.1763781537082192, 1.0803743403396568, 1.2532712203996637, 2.1775668731474704, -0.38381969994926934, 11.307665486706513, 21.272644292421518, 0.6060988006467033, -0.05066215249285122, 28.131265357172296, -2.3409812215471373, 0.011995020208835

### Итоговые значения средних абсолютных ошибок: 

* lassoModel = 5.984693119061619
* squaresModel = 6.231902929890508
* custom = 6.231902929890598

Кастомная реализация показала практически тот же результат, что и коробочные модели. Все модели имеют ошибку ~ 6 - 6.2 Учитывая, что среднее значение целевой переменной = 7.32 , результат мягко говоря не радует. Верояно были выбраны неподходящие признаки т.к. опираться на одну лишь корреляцию не совсем верно. Нужен более глубокий анализ, да и описание данных тоже не помешало бы т.к. не имея описания каждой колонки не совсем ясен смысл данных.